assets/js/components/SettingsNotice/SettingsNoticeWithIntersectionObserver.js

```diff
@@ -0,0 +1,70 @@
+/**
+ * Settings notice component.
+ *
+ * Site Kit by Google, Copyright 2021 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * External dependencies
+ */
+import { useIntersection } from 'react-use';
+import PropTypes from 'prop-types';
+
+/**
+ * WordPress dependencies
+ */
+import { useEffect, useRef, useState } from '@wordpress/element';
+
+/**
+ * Internal dependencies
+ */
+import SettingsNotice from './SettingsNotice';
+
+// TODO: This could renamed/moved to be more generic e.g. `ComponentWithIntersectionObserver`,
+// or refactored to a HOC.
+function SettingsNoticeWithIntersectionObserver( {
```

Comment:
As @techanvil mentioned [here](https://github.com/google/site-kit-wp/pull/8474#discussion_r1565687215), let's actually address this comment now and turn this into a more generic component, as this functionality doesn't need to be specific to the `SettingsNotice` component, I think. `ComponentWithIntersectionObserver` sounds great!

Reply:
@nfmohit @techanvil I have addressed this as a new `ComponentWithIntersectionObserver` component as per recent commits. I simply merely implemented it as a component with children, and in testing this seemed to work fine. Any concerns with this? See the commit at https://github.com/google/site-kit-wp/pull/8474/commits/32a7df0382754fe09410321799bcb4cb0da95986

Reply:
LGTM! I'll keep this open for @techanvil.

Reply:
Hey @10upsimon, thanks for addressing that. However, I notice that the approach you've taken introduces a new `div` element to wrap the children.

Seeing as the longer term intention here is to address the problem with `useIntersection()`, at which point we can remove this workaround, I'd rather we don't introduce additional markup via `ComponentWithIntersectionObserver` which will then be removed when the component is stripped out.

This can be addressed in quite a straightforward manner by rewriting the component as a HOC and passing the ref directly to the wrapped component.

I've had a quick go at this myself, creating `withIntersectionObserver()` and removing `ComponentWithIntersectionObserver`.

Although we usually cover our HOC functions with tests, I think we can skip it on this occasion, as long as we're clear about the fact it's just an interim measure.

On that note, it looks like this point in my previous [comment](https://github.com/google/site-kit-wp/pull/8474#discussion_r1565687215) has been missed:

> - Simon, if you haven't already, please can you file a followup issue to provide an improved version of useIntersection() and replace our current use of the version from react-use?

So, please can you do the following?
- Take a look at the HOC implementation that I've provided and see what you think, if you are happy with it I'll get this merged.
- File that GH issue to make sure we do follow up and address the issue with `useIntersection()`.

Cheers!

Cc @nfmohit

Reply:
Thanks @techanvil I took a look at your change, and while I need to understand it a little better, I get the gist of it and I tested it and the event still fires correctly, so I'm happy with those changes. 

I have indeed added it to my TODO list to create that issue, I did not realise the creation of said issue was a blocker for this PR to be approved, apologies for that, I'll get onto that soon. If we can move this ahead so long, that will help a lot.

Reply:
Thanks @10upsimon! As discussed on [Slack](https://10up.slack.com/archives/CBKKQEBR9/p1713355838010019), creating the followup isn't a blocker for this, I just wanted to make sure it's on your radar.

I'll get this one merged, cheers!